org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/codegen/BranchLabel.java

```diff
@@ -145,13 +168,16 @@ protected void trackStackDepth(boolean branch) {
 			if (this.targetStackDepth != this.codeStream.stackDepth) {
 				this.codeStream.classFile.referenceBinding.scope.problemReporter().operandStackSizeInappropriate(
 						this.codeStream.classFile.referenceBinding.scope.referenceContext);
-				if (this.targetStackDepth < this.codeStream.stackDepth)
+				if (this.targetStackDepth < this.codeStream.stackDepth) {
 					this.targetStackDepth = this.codeStream.stackDepth; // FWIW, pick the higher water mark.
+					this.operandStack = this.codeStream.operandStack.copy();
+				}
```

Comment:
This corrects an oversight unrelated to this defect. There are two pieces of state associated with operand type stacks. A scalar the stackDepth - no of words on stack - and a vector - the stack - holding the type bindings themselves. Existing code merely checks the scalars match at branch targets and if not chooses the bigger - we are FUBAR anyways - but we try to handle it as best as we can. In this process, we should also update the vector in sync. That it was not is an oversight and addressed here